### PR TITLE
Revert "Remove attribute restriction enforcing."

### PR DESCRIPTION
**Pull request by jodator**
This reverts commit 37e717ce

### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Fix: Restricted editing boundaries should not be crossed by attribute operations. See ckeditor/ckeditor5#5840.

---

### Additional information

* Follow-up to the https://github.com/ckeditor/ckeditor5-restricted-editing/pull/7

1.  :warning:  controversial one: Fixed by changing `range` for `AttributeOperation` in `applyOperation` method decorator. Nicer would be decorator for some method (`writer.removeAttribute()`?) since `operation.range` is "readonly". Fastest to implement.

---

It is a draft because it requires the previous one to be merged before and then this to be merged to `master` not the `i/5840-safe`.

### CODE REVIEW

**Comment by scofalik** [comment]
I have quite mixed feelings about this.

On one hand, I dislike this hack. On the other hand, it will be hard for me to convince you.

We try to avoid doing modifications on `applyOperation`. And for sure, modyfing read-only property of operation on the fly isn't something that I like. This is a rule we had to feel safe. Honestly, a good solution would shrink the range before an attribute operation is created.

I was wondering if we could handle it through a `Schema` check but this is a different scenario.

While ago, when reviewing track changes, I thought that it might be a good idea if command's `execute` would be split in two parts: one obtaining the ranges (by default from the current selection) and the other part would execute the action. This could be useful for track changes. It would be useful in this scenario too.

It is hard for me to tell where exactly the problem lies. I mean, where the fix should be applied. We have multiple choices. We cannot hook (at the moment, at least) into `model.setAttribute()` but we could hook into `LinkCommand.execute()`. But that would meant repeating a part of the logic. OTOH we do that in track changes (that's I wanted to do refactor in commands). Maybe we should treat restricted editing like track changes. :thinking:

Making writer functions a decorators was another way I wanted to refactor track changes. So I'd also +1 this solution.

Also, we should keep in mind, that at this moment, this is a problem just with a one command and it is a command which evaluates the range-to-change recklessly.

Finally, I was thinking what problems this hack may cause. One that I found was a scenario in collaboration, but you already said that the plugin is not meant to work in an integration where one collaborator is in "normal mode" and the other in "restricted mode".

Other than that and the fact that you are making a weird change in an unusual place and it may be hard to discover, I can't think of any problems. I'd rather see this in a post-fixer, to be honest, as it is the place to fix things like that. But I see myself that my argumentation is quite weak.

So, to sum up, I'd do this:

1.  Keep this hack.
2.  Go either with decorating `model.Writer` or start refactoring commands by splitting range evaluation and action execution into two (and have both decorated).
3.  Substitute this hack with a more correct solution, basing on decision made in point 2.

Alternatively, we might just split `LinkCommand` into a method that evaluates a range and a method that executes the action and "overwrite" the range-evaluating method.

**Comment by jodator** [comment]
> On one hand, I dislike this hack. On the other hand, it will be hard for me to convince you.

Nah, I don't like this hack either :D It is fugly. I just don't have faster and global way to alter all attribute operations as I could do easily with deleteContent or insert content.

> I'd rather see this in a post-fixer

Yeah. I don't know why I didn't go with post-fixer in the first place. Probably I was too anchored with a solution for inserting/deleting conentent.

**Comment by jodator** [comment]
@Reinmar I've changed the solution to a post fixer - it can be added to the other PR as a post-fixer looks safe. But we can merge it quickly after the other PR.

File: src/restrictededitingmodeediting.js [hunk 1]
```diff
@@ -386,3 +362,74 @@ function isRangeInsideSingleMarker( editor, range ) {
...
+					}
+				}
+			}
+		}
+
+		return changeApplied;
+	};
+}
+
+// Calculates a difference of one range with many ranges (exclusion).
```
Comment (scofalik) [inline src/restrictededitingmodeediting.js]: We already have such function in track changes :(. I guess it is time to move it to `model.Range`.

**Comment by jodator** [comment]
Since no-one complains I'm merging this PR.
